### DISCUSSION TITLE
Global observables

### DISCUSSION THREAD

**Discussion by wverkerke** [hep-statistics-standard/hep-statistics-standard#20] [Standard]
I leave the following short note on 'global observables' here in this issue track on the request of Carsten.

**Global observables**

Many LHC measurements rely on information that is not measured from the dataset that is analysed, but that is externally provided. Examples of such information range from parameters specifying efficiencies and resolutions related to detector specifications,  to parameters specifying theoretical information such as cross-sections and associated uncertainties. As nominal values of the  required external information (i.e. nominal cross-sections, nominal efficiencies) are typically directly encoded in the likelihood function of the measurement,  the role of the parameters is mostly to specify the uncertainties on this external information.  

The canonical way in which this information is encoded in the likelihood function of the physics data, is that the quantity subject to an uncertainty is represented by a response function instead of a constant,  e.g. for an uncertain background event count B from simulation, an example response function could look like B(alpha_qcd, alpha_pdf) = B(1 + 0.10*alpha_qcd + 0.05*alpha_pdf),  encoding a 10% uncertainty due to 'QCD' and a 5% uncertainty due to 'PDF'.  In this form, all of the 'physics' information on the uncertainty quantity B is encoded in the response function (nominal value of B and (relative) magnitude of each source of uncertainty), however the physics likelihood must still be complemented with 'subsidiary' likelihood functions that 'measure' the externally provided information alpha_qcd and alpha_pdf. Parameters representing additive uncertainties, which are most common, like alpha_qcd and alpha_pdf in the example, are typically modeled with unit Gaussian p.d.f.s , G(alpha_obs==0|alpha_qcd,1), as they agree with the convention encoded in the response model: that value zero represents the nominal value, and that a variation of +/-1 represents one standard deviation. 

Note that subsidiary measurements for a parameter alpha are formulated as a likelihood in alpha, i.e. G(0,alpha,1), and not as a constraint on random variable alpha G(alpha,0,1), as in the frequentist paradigm alpha cannot be treated as a random variable. This choice of formulation for the external information as 'measurement' rather than as 'constraint' allows for the interpretation of the likelihood in both the frequentist and bayesian framework, but implies that there is (potentially large) set of 'subsidiary measurements', which all have value zero, that are part of the 'data' of any measurement. 

A 'business as usual' approach to such models, i.e. treating subsidiary data like any other observed data from the analysis would imply that the measured dataset D(x) is supplemented with all  required subsidiary observables, i.e. D(x,obs_alpha0,obs_alpha1,.... obs)alphaN), which can then be interpreted by a pdf of the form F(x)*G1(obs_alpha1)*...*Gn(obs_alphaN). In practice this is rather cumbersome as low-dimensional dataset in observable(s) x become high-dimensional datasets (with realistically up to 1000 alpha observables), that contain trivial data (all zeros), which also do not represent any tangible information (since that is encoded in the response function in the likelihood function).

To avoid this practical complication, the notion of Global Observables was introduced in ROOT/RooFit/RooStats. A global observable is defined as an observable whose value is not encoded in the dataset D, but rather in the PDF. In other words, instead of D(x,alpha1=0...alphaN=0) and F(x,alpha1...alphaN), the information is represented as D(x) and F(x,alpha1=0....alphaN=0). All RooFit models already include a representation of all observables, hence a (default) value for all observables is already there. The effect of declaring an observable 'Global' in a usage context (i.e. ML fit of a pdf to data), has the effect of the likelihood taking the observable value from the internal representation in the pdf, rather than from the provided dataset, thereby sidestepping the requirement of having to provide the values in the dataset.

**Comment by cburgard** [comment]
Thanks a lot @wverkerke for posting this explanation!

The concept of "Global Observables" is easily reflected in RooFit by virtue of the "Constraint" Pdfs not being linked to any "regular" observable present in the dataset, but to some (constant) variable in the workspace. The identity of these global observables are then communicated e.g. to the ModelConfig via the special list "GlobalObservables".

In HS3, we currently don't have a clear reflection of this concept.
I can see several ways forward:
 1. we neglect the presence of "Global Observables" and treat them as regular observables, creating single-point datasets for each of them. Pro: This is extremely explicit and understandable to everyone looking at the file. Con: We will end up with thousands of meaningless datasets in every file.
 2. we group all of them into a special magical dataset comprised of only 0s/1s. Pro: Still pretty explicit, but much more compressed. Con: Translation back and forth from RooFit might be non-trivial, and special types of Pdfs might be needed to encode this ("list of unit gaussians")
 3. we do exactly what RooFit does and declare a "special list" of variables of some sort, e.g. by allowing for another top-level component that could be called "implicit_data" or indeed "global_observables", that is just a map of variable names to their respective default values. Pro: Maps directly to the RooFit implementation. Con: Maybe not easy to understand for people enocuntering the concept for the first time.
 4. we encode the implicit data through the usage of constraints, by delaring that specific constraint types "always" map to implicit data of a specific type. Pro: extremely compact, in-line with what pyhf does already. Con: Not sure how well this generalizes. Difficult to use for some procedures such as throwing toys (which need to include alternative values for the global observables).

I'd love to hear everyones thoughts on this issue!

**Comment by oschulz** [comment]
What was the approach we did discuss with @lukasheinrich again?

**Comment by cburgard** [comment]
I think what we discussed with @lukasheinrich is closest to option (2) from the list above. 
I would love to get some comments, ideally by @wverkerke, @lmoneta and @guitargeek, on how they would evaluate the "Con" argument from above, and if any other issues are to be expected with this approach.

What is currently implemented is basically (4), but I'm pretty sure we are going to cause problems when pursuing this option any further, so we will probably want to change.

The main question is if any of the other options (1) and (3) would be a better match for us, or if (2) is (as we previously discussed) indeed the "least bad" option.

Any opinions are highly welcome!

**Comment by lukasheinrich** [comment]
Hi all, (also adding @cranmer in case he has a comment)

I think there is a good middle ground where the subsidiary data is treated as "business as usual" data without completely cluttering the likelihood function with many terms if you allow for "array" (diagonal) pdfs a la

`DiagonalGauss([d1,d2,d3....,dn] | [µ1,µ2,µ3,...µn], [σ1, σ2, σ3, ... σn]) := G(d1|µ1,σ1)*G(d2|µ2,σ2)...G(dn|µn,σn)`


`DiagonalPoisson([d1,d2,d3....,dn] | [λ1,λ2,λ3,...λn]) := Pois(d1|λ1)*Pois(d2|λ2)...Pois(dn|λn)`


even 100's of subsidiary measurements are then expressible in a very compact way and this is what we do practically in `pyhf`

there are three components in the likelihood
* the main observation model 
* the Gaussiuan subsidiary likelihoods  (possibly many in a "diagonal" array form)
* the Poisson subsidiary likelihoods (possibly many in a "diagonal" array form)

**Comment by nsmith-** [comment]
I think "array observations" are going to be popular in general. Even for regular observables, we have a concept of `DiagonalPoisson` for 1d binned datasets. So I would +1 on having explicit array observables for the constraint terms. Also, we may want to store sampled frequentist toys as "observations" in which case it is much more conceptually straightforward to have one "toy observation" include all observable values.

**Comment by cburgard** [comment]
@lukasheinrich thanks a lot for the follow-up! As a minor point, there's also log-normal as a constraint type I think, but that's a trivial extension.
More broadly, one could think about a syntax which allows to map any arbitrary 1d-distribution to an array of observations, but perhaps that's going too far, since the distinct types of diagonal measurements might allow for framework-specific optimizations, which we might loose if we go with generic array mappings (also, this would open the whole pandoras box of functors, which I'm not ready to dive into yet).
So I think supporting array constraints is a good middle ground!

Perhaps @guitargeek can comment, but it was my understanding that also on the ROOT side we were hoping to create summary classes for the constraint terms, which would make this mapping infinitely easier.

**Comment by jonas-eschle** [comment]
That's an interesting consideration, and it just becomes relevant with many parameters and surely needs to be addressed.

In general, I would lean towards 1, but mixed with what lukas proposed: using diagonal PDFs.

Let me drop two more flavours in the discussion:
- constraints can in general be arbitrary (I think the focus on Gauss and Poisson (+ LogNormal) comes a lot from templated, binned fits, so I'd definitely not specialise on specific constraints, but then again, the DiagGauss etc will directly solve the problem.
- values of 0 and 1 seem again often used in this templated, binned fits. While we can always transform (theoretically) to a gaussion, practically I see a lot of other numbers, i.e. "mass and width of particle" as param + constraint. So 0s and 1s are a crucial case, but not the only one.

So advocating for 1:
- the names can be important _for combinations_ (as also discussed in https://github.com/hep-statistics-serialization-standard/hep-statistics-serialization-standard/issues/10). Each value has a clear identifier.
- For the sake of generality, I would tend to not introduce any new concept/special keywords _if not really necessary_. As is,  we can implement all of it (assuming DiagGauss or just NDim Gauss in general). If this really becomes an issue, practically, then we can still see _as an alternative method_ to provide shortcuts, but the general way will always be possible.
- maybe we have 1000 names, maybe a million. This is an ndim dataset. So it's a list of names, a list of numbers. Names are needed, numbers aren't an issue either. So be it, that's why we create the standard. Yes, human readable is part of it, but that still seems very readable, as much as 1k values are regarded as "human readable". There is nothing agains having a converter to display something differently (and we still have the pyhf-format that can be converted to and from, ideally).

**Comment by cburgard** [comment]
I've just had a discussion with @guitargeek, which I'm paraphrasing now:

Technically, we probably want to support both options 1 and 2.
Option 1 has the benefit of being super explicit, and for small setups (think "tutorial"), it's probably easy. Plus, not every framework will implement the "collective constraint terms" from the get-go.
Option 2 does seem like the way to go for large setups. I'm just lining out here how these things would look as a proposal, feel free to disagree and post suggestions.

Option 1
===
```
{
  "likelihoods": [
    {
      "name": "main_lh",
      "distributions": ["measurement_dist", "aux_measurement_1_dist", "aux_measurement_2_dist", /*...*/],
      "data": ["measurement", "aux_measurement_1", "aux_measurement_2", /*...*/]
    }
  ],
  "distributions": [
    {
      "name": "measurement_dist",
      // this is the pdf that actually describes your measurement
    },
    {
      "name": "aux_measurement_dist_1",
      "type": "normal_dist",
      "mu": "np_1",
      "x": "nom_np_1",
      "sigma": 1.0
    },
    {
      "name": "aux_measurement_dist_2",
      "type": "poisson_dist",
      "mu": "np_2",
      "x": "nom_np_2"
    },
    /*...*/
  ],
  "data": [
    {
      "name": "measurement",
      // this is where the actual measurement goes
    },
    {
      "name": "aux_measurement_1",
      "type": "point",
      "value": 0,
      "axes": [
        {
          "x": "nom_np_1"
        }
      ]
    },
    {
      "name": "aux_measurement_2",
      "type": "point",
      "value": 100,
      "axes": [
        {
          "x": "nom_np_2"
        }
      ]
    },
    /*...*/
  ]
}

```

Option 2
===
```
{
  "likelihoods": [
    {
      "name": "main_lh",
      "distributions": ["measurement_dist", "poisson_constraints", "gauss_constraints" ],
      "data": ["measurement", "poisson_aux_data", "gauss_aux_data", /*...*/]
    }
  ],
  "distributions": [
    {
      "name": "measurement_dist",
      // this is the pdf that actually describes your measurement
    },
    {
      "name": "gauss_constraints",
      "type": "diagonal_gauss_dist",
      "mu": ["np_1", "np_2", /*...*/],
      "x": ["nom_np_1", "nom_np_2", /*...*/],
      "sigma": 1.0
    },
    {
      "name": "poisson_constraints,
      "type": "diagonal_poisson_dist",
      "mu":  ["np_101", "np_102", /*...*/],
      "x": ["nom_np_101", "nom_np_102", /*...*/],
    },
    /*...*/
  ],
  "data": [
    {
      "name": "measurement",
      // this is where the actual measurement goes
    },
    {
      "name": "aux_gauss_measurement",
      "type": "point",
      "value": [0, 0, /*...*/],
      "axes": [
        {
          "x": "nom_np_1"
        },
        {
          "x": "nom_np_2"
        },
       /*...*/
      ]
    },
    {
      "name": "aux_poisson_measurement",
      "type": "point",
      "value": [100, 100, /*...*/],
      "axes": [
        {
          "x": "nom_np_101"
        },
        {
          "x": "nom_np_102"
        },
        /*...*/
      ]
    },
    /*...*/
  ]
}
```

**Reply by jonas-eschle** [reply]
Thanks for the example, I may misunderstood the difference between 1 and 2. To me, they look the same and should both be valid in our standard; in fact, not having one implies not having the other. Having 2 in seems to me to be a matter of having diagonal PDFs in.

My main point is that 2 adds more models that do not really add to the statistical expression. I am less concerned about size and readability of larger models, as a) if size is not an issue, it's fine, and if it is, we need to come up with other things (as for unbinned data) and b) "human-readable" becomes relative for large models anyway and the standard becomes more "machine-readable" in ASCII :)
No strong opinion though, having 2 available or not seems rather a question of having diagonal PDFs in or not.

**Reply by nsmith-** [reply]
I agree with @jonas-eschle here, it seems option 1 and 2 are both the same as far as how the likelihood semantics work. The addition of array observables is something I support (as mentioned above) but unrelated to the question of where auxiliary constraints go.

**Comment by cburgard** [comment]
Since nobody objects, I assume that supporting both Option 1 and Option 2 is accepted.

**Comment by cburgard** [comment]
To follow up on this, we currently have this as an example in our standard

```
"likelihoods":[
   {
      "name":"likelihood1",
      "distributions":[
         "dist1",
         "dist2",
         "single_dimensional_dist_1",
         ...
      ],
      "data":[
         "data1",
         "data2",
         0,
         ...
      ],
      "aux_distributions" : [
         "regularization_term"
      ]
   },
   ...
]
```

This is also pretty charming, because it would allow to give point-data in-place.

We should make up our mind if we want to keep that option, or if we believe that this is unnecessary and leads to confusion.
I think a point can be made that this way, we would allow for "unnamed" or "anonymous" data, which might cause problems because it is not entirely clear what variable this data belongs to. We do, however, in our standard currently state this:

> Alternatively, the data-values for single-dimensional distributions can be given in-line. For example, this can be used for constraint terms representing auxiliary measurements.

If we really want to pursue this path, we could also, instead of explicitly having `aux_distributions`, allow for auxiliary distributions to be added to the regular distributions list. This could even include things like regularization terms, if we decide to allow `None` as a valid entry in the data-list.

How do people feel about this?

I personally like the explicitness of `aux_distributions` and dedicated data-points, and would thus have a slight preference for disallowing anonymous data. But other people may well have very different opinions, and I seem to remember that the capacity for anonymous data was added on suggestion of @oschulz

**Reply by jonas-eschle** [reply]
I agree with your concern and I share the preference of a dedicated datapoint.

> Alternatively, the data-values for single-dimensional distributions can be given in-line. For example, this can be used for constraint terms representing auxiliary measurements.

Seems to me indeed dangerous, as it's not generally well defined (think of combinations). Rather strict and clear seems better to me, in the end, these files are mostly written by machines, so why not be explicit about it

**Comment by will-cern** [comment]
At Carsten's request, I will make a couple of comments here.

Since Wouter opened this discussion, the ability to store a list of global observable values in the RooFit dataset was added. I strongly believe that it makes sense that the dataset be thought of as having a list of global observable values, but it was interesting to hear from Wouter the original reasons why they decided against having datasets hold global obs values (a lot of them being zeros). Global observables being part of datasets makes toy dataset generation more coherent, since global obs should be randomized in a toy as well, so we need to know what are the global obs values to go along with the regular obs in the toy. 

I would be interested to here what are the objections today to have a list of global observable values in the dataset object of HS3?

**Reply by cburgard** [reply]
To me, the main objection is that it introduces a non-trivial concept to an otherwise trivial object.
HS3 knows several types of datasets already (binned and unbinned). Following your line of thinking, we would want to augment both of them with additional "global" observables, that would act as "decorators". 
So if I have a datset with 1000 entries in the variables a, b and c, and I would now add the global observable d, I would have 1000 values of a,b,c, and then one value in d. This is functionally identical to having one dataset that has 1000 entries in the variables a, b and c, and then another dataset with one entry in the variable d.

To me, it's not clear what we would gain from being able to merge datasets in the way you describe. My gut feeling is that we would fare better by just allowing a concept of combined datasets, where we would now define the first dataset with 1000 entries in a, b and c, the second dataset with one entry in d, and a third "combined" dataset that just is the combination of the first and the second. This way, we would gain more flexibility, because we can now also combine datasets from different channels (one each) and for global observables (one each) in a way that is free of conflicts and overlaps. Otherwise, we would need to decide what we do if we combine two channels that have a non-congruent set of global observables.

But perhaps there's a strong motivation to integrate the global observables in the primitive dataset types that I don't see at the moment?

**Reply by will-cern** [reply]
How do you signify how the join between two datasets is supposed to happen? e.g. joining two datasets with the same number of entries (a horizontal concatenation) vs what we are discussing here with one ds with one entry and another with many entries? How do you make it clear for the extended term of the NLL that the number of entries is determined by the 'many entries' dataset? Even if that dataset has 0 entries in it? 
Not saying this isn't all possible to sort out, but I guess I'm thinking I'd rather first see this concept implemented in RooFit itself before it made its way into HS3. We have an ability to "Link" datasets together in RooFit, but that is a vertical concatenation with an additional discrete observable to label the child dataset. It sounds like we need to have horizontal linking for what you want. And it needs to be smart ..... e.g. if you have a ds1 for the robs of chan1, ds2 for the globs of chan1, ds3 for the robs of chan2, ds4 for the globs of chan2. You want to horizontally join ds1 and ds2 to make the combined ds for chan1, and similar for ds3 and ds4. And then if you want to combine those two channel-datasets? That's a vertical join, yes? But not of the global observables ... it's a merge of those .... so it has to know the right thing to do. The logic feels complex to me at the moment, but perhaps I've overcomplicated it?

**Reply by cburgard** [reply]
In my (perhaps naive) idea, we would always just have a vertical combination, that is, we would require datasets that are supposed to be combined to have no overlapping variables, and the resulting combined dataset would always be the "product" of the original ones. This way, each constraint term would basically become its own channel (in RooFit-speak). 
But perhaps there is some important capability that we lose by doing this - feel free to point me to it!

**Reply by will-cern** [reply]
What happens if the channels share observables? This will surely happen for the global observables, yes? So then your combined dataset would have two entries for those observables, and then the constraint pdfs would be evaluated twice, so the contraint would be incorrectly doubled in strength?
